zstd-safe/src/lib.rs

```diff
@@ -1048,15 +1084,129 @@ pub struct InBuffer<'a> {
     pub pos: usize,
 }
 
+/// Describe a resizeable bytes container like `Vec<u8>`.
+///
+/// Can start from uninitialized memory, and will be partially filled.
+///
+/// Should be implemented by a contiguous chunk of memory.
+///
+/// The main implementors are:
+/// * `Vec<u8>` and similar structures. These can start empty with a non-zero capacity, and they
+///   will be resized to cover the data written.
+/// * `[u8]` and `[u8; N]`. These must start already-initialized, and will not be resized. It will
+///   be up to the caller to only use the part that was written.
+pub trait WriteBuf {
```

Comment:
I think this trait might need to be `unsafe` since `as_slice()` needs to uphold the initialization invariant but probably shouldn't itself be an unsafe method.

Reply:
I agree it needs to be an unsafe trait, but for different reasons: that `capacity()` must not lie about the actual length of available data.

`as_slice` itself, if implemented safely, cannot return uninitialized memory (if it could, then that uninitialized memory would probably be visible without this method). We're not _relying_ on what this implementation provides. `filled_until` is the main one that needs to be properly implemented: it's both unsafe to use _and_ to implement (maybe that should also justify unsafe trait?)

Reply:
Ahh, yeah this makes sense.